### PR TITLE
Add a Fluid-Fluid module

### PR DESCRIPTION
**Pull request by MakisH**
This Pull Request adds a simple Fluid-Fluid module.

I have currently implemented reading and writing pressures and velocities, but I have not yet tested it.

See issue #60.

### CODE REVIEW

File: tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    location    "constant";
+    object      transportProperties;
+}
+// * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * //
+
+transportModel  Newtonian;
+
+nu              nu [ 0 2 -1 0 0 0 0 ] 1e1;
+
+rho             rho [1 -3 0 0 0 0 0] 1e-0;
```
Comment (MakisH) [inline tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties:15]: We can probably move the density into `system/preciceDict`

File: tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0"?>
+
+<precice-configuration>
+
+    <log>
+        <sink type="stream" output="stdout" filter="%Severity% > debug" enabled="true"/>
```
Comment (MakisH) [inline tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml:6]: Here we need to add the usual `---[precice]` format.

File: FF/FF.C [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+(
+    const Foam::fvMesh& mesh
+)
+:
+mesh_(mesh)
+{}
+
+bool preciceAdapter::FF::FluidFluid::configure(const IOdictionary& adapterConfig)
+{
+    DEBUG(adapterInfo("Configuring the CHT module..."));
```
Comment (MakisH) [inline FF/FF.C:17]:
```suggestion
    DEBUG(adapterInfo("Configuring the FF module..."));
```

File: FF/FF.C [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    DEBUG(adapterInfo("    user-defined solver type : " + solverType_));
+    
+    // Read the name of the velocity field (if different)
+    nameU_ = FFdict.lookupOrDefault<word>("nameU", "U");
+    DEBUG(adapterInfo("    velocity field name : " + nameU_));
+
+    // Read the name of the pressure field (if different)
+    nameP_ = FFdict.lookupOrDefault<word>("nameP", "p");
+    DEBUG(adapterInfo("    pressure field name : " + nameP_));
+
```
Comment (MakisH) [inline FF/FF.C:63]: Here we (may) need to also read density from `system/preciceDict`.

File: FF/FF.C [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+std::string preciceAdapter::FF::FluidFluid::determineSolverType()
+{
+    // NOTE: When coupling a different variable, you may want to
+    // add more cases here. Or you may provide the solverType in the config.
+
+    std::string solverType;
+    
+    // TODO: Implement properly
+    solverType = "incompressible";
```
Comment (MakisH) [inline FF/FF.C:75]: We can reuse the implementation from #124.

File: tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties (hunk 1 above)
Comment (MakisH) [reply tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties]: We actually don't need the density here. I removed it.

File: tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml (hunk 2 above)
Comment (MakisH) [reply tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml]: Fixed. We also needed to restrict output to rank0.

File: FF/FF.C (hunk 4 above)
Comment (MakisH) [reply FF/FF.C:62]: We don't need the density here at the moment.

File: FF/FF.C (hunk 5 above)
Comment (MakisH) [reply FF/FF.C]: Implemented.

File: tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    <m2n:sockets from="Fluid1" to="Fluid2"/>
+
+    <coupling-scheme:serial-implicit>
+        <time-window-size value="0.01"/>
+        <max-time value="1.0"/>
+        <participants first="Fluid1" second="Fluid2"/>
+        <exchange data="Velocity" mesh="Fluid1-Mesh" from="Fluid1" to="Fluid2"/>
+        <exchange data="PressureGradient" mesh="Fluid1-Mesh" from="Fluid1" to="Fluid2"/>
+        <exchange data="Pressure" mesh="Fluid2-Mesh" from="Fluid2" to="Fluid1"/>
+        <!--<exchange data="VelocityGradient" mesh="Fluid2-Mesh" from="Fluid2" to="Fluid1"/>-->
```
Comment (MakisH) [inline tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml:60]: This combination diverges very early in the simulation.

File: tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties [hunk 7]
```diff
@@ -12,6 +12,4 @@ transportModel  Newtonian;
 
 nu              nu [ 0 2 -1 0 0 0 0 ] 1e1;
 
-rho             rho [1 -3 0 0 0 0 0] 1e-0;
```
Comment (pachesp) [inline tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties]: (testing comment) rho is in nu

File: FF/FF.C [hunk 8]
```diff
@@ -14,7 +14,7 @@ mesh_(mesh)
 
 bool preciceAdapter::FF::FluidFluid::configure(const IOdictionary& adapterConfig)
 {
-    DEBUG(adapterInfo("Configuring the CHT module..."));
+    DEBUG(adapterInfo("Configuring the FF module..."));
```
Comment (pachesp) [inline FF/FF.C:17]: well spotted

File: FF/FF.C [hunk 9]
```diff
@@ -69,11 +68,31 @@ std::string preciceAdapter::FF::FluidFluid::determineSolverType()
...
+
+    if (mesh_.foundObject<volScalarField>("p"))
+    {
+      volScalarField p_ = mesh_.lookupObject<volScalarField>("p");
+
+      if (p_.dimensions() == pressureDimensionsCompressible)
+        solverType = "compressible";
+      else if (p_.dimensions() == pressureDimensionsIncompressible)
+        solverType = "incompressible";
+    }
```
Comment (pachesp) [inline FF/FF.C:84]: I am unfamiliar with the units for compressible and incompressible pressure, but so far I have been using an incompressible solver, and my pressure units are kgm/s^2 m^2 (1, -1, -2, 0, 0, 0, 0), isnt this the other way?

File: tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties (hunk 7 above)
Comment (MakisH) [reply tutorials/FF/pipe-pipe/pimpleFoam-pimpleFoam/Fluid1/constant/transportProperties]: So, in any case, we don't need it anywhere at the moment, right?

File: FF/FF.C (hunk 9 above)
Comment (MakisH) [reply FF/FF.C:84]: In an incompressible solver, OpenFOAM [uses the kinematic pressure](https://www.openfoam.com/documentation/guides/latest/doc/guide-applications-solvers-variable-transform-kinematic-pressure.html), which is in `m^2 / s^2` and in named `p`. Compressible solvers use the static pressure, which is in  `Pa` or `kg / (m s^2) and is also named `p`.

You are probably talking about interFoam, which [splits the pressure](https://www.openfoam.com/documentation/guides/latest/doc/guide-applications-solvers-multiphase-interFoam.html) into `p` and `p_rgh`, both in `Pa`. We should indeed consider this as a special case when we add support for interFoam, good point!

Comment (MakisH) [reply FF/FF.C:84]: In any case, this seems to correctly distinguish between `pimpleFoam` (incompressible) and `sonicLiquidFoam` (compressible).

Comment (pachesp) [reply FF/FF.C:84]: cool, then in case for interFoam, lets lets a `TODO` there

Comment (MakisH) [reply FF/FF.C]: I quickly tried with interFoam, it classifies it as an incompressible solver. We currently don't use the solverType anyway, but I added a TODO. We will probably need it in the future.

File: tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        <max-iterations value="100"/>
+        <relative-convergence-measure limit="1.0e-3" data="Pressure" mesh="Fluid2-Mesh"/>
+        <!--<relative-convergence-measure limit="1.0e-6" data="PressureGradient" mesh="Fluid2-Mesh"/>-->
```
Comment (uekerman) [inline tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml:64]:
```suggestion
```

File: tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml (hunk 6 above)
Comment (uekerman) [reply tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml:60]: The problem was the too weak convergence of OpenFOAM combined with the too strict convergence measures of preCICE.

Comment (MakisH) [reply tutorials/FF/pipe-pipe/sonicLiquidFoam-sonicLiquidFoam/precice-config.xml:60]: Oh, I was working on this already, thanks!

**Comment by MakisH** [comment]
@uekerman In my local copy I had cleaned up the case files, I will clean them again.

There are indeed many whitespace changes, but I plan to work on #126 soon afterwards, as right now out formatting is quite inconsistent and we will have similar problems later on.

**Comment by MakisH** [comment]
This is now ready enough to merge to develop, so that we are able to built upon it. Thank you both for the reviews!

**Comment by FrankyOpenFOAM** [comment]
Hello, in system file, there is a  "precicedict". In precicedict, there is a interface, which is used to determine patch to transfer the data.    But in my work, I want to transfer all of my data in whole computational domain to other openfoam solver.  Do you know if there is a way to do this?

**Comment by MakisH** [comment]
What you want to do is what we refer to as "volume coupling". A few people have tried this already (e.g., #183, #182, #181) and we are working to implement this properly and fully in the adapter.

**Pull request merged**